connector_amazon/models/amazon_sale_importer.py

```diff
@@ -3,14 +3,204 @@
+        ]
+        file = StringIO.StringIO()
+        file.write(report)
+        file.seek(0)
+        reader = unicodecsv.DictReader(
+            file, fieldnames=fieldnames, delimiter='\t', quoting=False,
+            encoding=meta_attachment.amazon_backend_id.encoding)
+        reader.next()  # we pass the file header
+        sales = self._extract_infos(reader)
+        file.close()
+        return self._create_sales(sales, meta_attachment)
+
+    def _extract_infos(self, reader):
+        sales = {}
+
+        def reset_empty(value):
+            if value == '':
+                return False
+            return value
+
+        for line in reader:
+            if not line.get('order-item-id'):
+                continue
+            if line['order-id'] in sales:
+                vals = sales[line['order-id']]
+                vals['lines'].append(self._get_sale_line(line))
+            else:
+                vals = {
+                    'sale': {
+                        'origin': line['order-id'],
+                        'date_order': line['purchase-date'],
+                    },
+                    'partner': {
+                        'email': line['buyer-email'],
+                        'name': line['buyer-name'],
+                        'phone': line['buyer-phone-number'],
+                    },
+                    'part_ship': {
+                        'name': line['recipient-name'],
+                        'phone': reset_empty(line['ship-phone-number']),
+                        'street': reset_empty(line['ship-address-1']),
+                        'street2': reset_empty(line['ship-address-2']),
+                        'street3': reset_empty(line['ship-address-3']),
+                        'city': line['ship-city'],
+                        'state': reset_empty(line['ship-state']),
+                        'zip': reset_empty(line['ship-postal-code']),
+                        'country': line['ship-country'],
+                    },
+                    'lines': [self._get_sale_line(line)],
+                }
+            sales[line['order-id']] = vals
+        return sales
+
+    def _get_sale_line(self, line):
+        return {
+            'item': line['order-item-id'],
+            'sku': line['sku'],
+            'name': '[%s] %s' % (line['sku'], line['product-name']),
+            'product_uom_qty': line['quantity-purchased'],
+            # price is in full tax, vat is computed in odoo
+            'price_unit': float(line['item-price']) + float(line['item-tax']),
+            'shipping': float(line['shipping-price']) + \
+            float(line['shipping-tax']),
+        }
+
+    def _create_sales(self, sales, meta_attachment):
+        """ We process all sale orders of the file
+        """
+        backend = meta_attachment.amazon_backend_id
+        for order_item, sale in sales.items():
+            partner, part_ship = self._get_partners(sale)
```

Comment:
Maybe it will be cleaner to have somthing like

```
partner = self._get_existing_partner(sale['partner'])
if partner:
   prepare partner vals
   update partner
else:
   prepare partner vals
   update partner

address= self._get_existing_address(partner, sale['part_ship'])

if not address:
    prepare address (the match of state is done in the _prepare_address method)
    create address
```